src/test/java/seedu/address/ui/PersonListPanelTest.java

```
@@ -48,11 +51,63 @@ public void display() {
 
     @Test
     public void handleJumpToListRequestEvent() {
+        setUp(TYPICAL_PERSONS);
         postNow(JUMP_TO_SECOND_EVENT);
         guiRobot.pauseForHuman();
 
-        PersonCardHandle expectedCard = personListPanelHandle.getPersonCardHandle(INDEX_SECOND_PERSON.getZeroBased());
-        PersonCardHandle selectedCard = personListPanelHandle.getHandleToSelectedCard();
-        assertCardEquals(expectedCard, selectedCard);
+        PersonCardHandle expectedPerson = personListPanelHandle.getPersonCardHandle(INDEX_SECOND_PERSON.getZeroBased());
+        PersonCardHandle selectedPerson = personListPanelHandle.getHandleToSelectedCard();
+        assertCardEquals(expectedPerson, selectedPerson);
+    }
+
+    @Test
+    public void performanceTest() throws Exception {
+        File xmlFile = createManyPersonsXmlFile();
+        XmlSerializableAddressBook xmlAddressBook =
+                XmlUtil.getDataFromFile(xmlFile, XmlSerializableAddressBook.class);
+        ObservableList<Person> personList =
+                FXCollections.observableArrayList(xmlAddressBook.toModelType().getPersonList());
+
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        setUp(personList);
+        guiRobot.interact(personList::clear);
```

Just a comment, if we did not apply the fix, then the test will just continue to run even if it has already exceeded 2 seconds (in fact, when I tried this on `master`, the window just froze, and I waited patiently for ~30 seconds before manually aborting it).

Is there a way to fail faster? (i.e. timeout on 2 seconds, rather than waiting for `personList::clear` to finish.)

---

Hmm, 1 way I can think of is multithreading: 1 thread on the timer, 1 thread on clearing the list. So once the timer is up, it will check whether the list is empty. If not, fail the test. Sounds ok?

---

> Hmm, 1 way I can think of is multithreading: 1 thread on the timer, 1 thread on clearing the list. So once the timer is up, it will check whether the list is empty. If not, fail the test. Sounds ok?

You can give it a try, I seem to recall writing a similar method to test methods with a timeout stopwatch though, was that no longer in the codebase or...?

---

Yup, we have `GuiRobot#waitForEvent(BooleanSupplier, int)` implemented by you :P However, currently, whenever this method is called, the concurrency is automatically handled. For e.g. when we are waiting for the browser to load, we simply call this method because the loading of the browser is automatically handled on another thread, so we can call `waitForEvent(...)` and let it block the current thread. (see `WebViewUtil#waitUntilBrowserLoaded(BrowserPanelHandle)`)

In this case, we can't do this:
```
guiRobot.waitForEvent(personList::isEmpty, 2000); // BLOCK :( so this definitely fails
initUi(personList);
guiRobot.interact(personList::clear);
```
nor this:
```
initUi(personList);
guiRobot.interact(personList::clear); // method only returns after personList is cleared
guiRobot.waitForEvent(personList::isEmpty, 2000); // this method definitely passes
```

So we have to manually perform the multi-threading.